File: 中级算法/字谜分组/C++/original.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> result;
        vector<string> temp;
        string s;
        int flag;
        for(int i=0; i<strs.size(); i++){
            s=strs[i];
            sort(s.begin(), s.end());
            flag=0;
            for(int j=0; j<temp.size(); j++){
                if(temp[j]==s){
                    cout<<i;
                    result[j].push_back(strs[i]);
                    flag = 1;
                    break;
                }
            }
            if(flag==0){
                temp.push_back(s);
                result.push_back({strs[i]});
            }
        }
        return result;
    }
};
/* 思路：用一个vector temp存放相同字母的词，遍历整个strs，把元素排序，遍历temp，
 * 判断temp中是否有该相同字母的词，如果有，把这个元素加入result中对应的vector中，
 * 如果没有，用一个flag标记，把这个元素的原型和排序后的字符串分别加入result和temp中，
 * 返回result。
 */

File: 中级算法/数组和字符串/字谜分组/python/original.py
class Solution:
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        思路：用一个字典建立索引表，键为同字母的词，
        键值为该字母的列表在返回结果中的位置index，
        遍历strs，先把元素排序并转化成元组，
        如果该元组没在字典dic中，把这个元素以列表
        的形式加入result，把元组变成字典键，对应
        的键值为该元素在result中的index，如果该元组在
        字典中，把该元素加入到result中对应的列表中，
        返回result。
        """
        flag = 0
        dic = {}
        result = []
        for i in strs:
            j = tuple(sorted(i))
            if j not in dic:
                dic[j]=flag
                flag+=1
                result.append([i])
            else:
                result[dic[j]].append(i)
        return result